### PR TITLE
[SPARK-3650][GraphX] Triangle Count handles reverse edges incorrectly

### PR DESCRIPTION
**Pull request by insidedctm**
@jegonzal @ankurdave please could you review
## What changes were proposed in this pull request?

Reworking of @jegonzal PR #2495 to address the issue identified in SPARK-3650. Code amended to use the convertToCanonicalEdges method. 
## How was the this patch tested?

Patch was tested using the unit tests created in PR #2495

### CODE REVIEW

File: graphx/src/main/scala/org/apache/spark/graphx/lib/TriangleCount.scala [hunk 1]
```diff
@@ -80,15 +103,18 @@ object TriangleCount {
...
-        val dblCount = optCounter.getOrElse(0)
-        // double count should be even (divisible by two)
-        assert((dblCount & 1) == 0)
-        dblCount / 2
+    graph.outerJoinVertices(counters) { (vid, _, optCounter: Option[Int]) =>
+      val dblCount = optCounter.getOrElse(0)
+      // This algorithm double counts each triangle so the final count should be even
+      val isEven = (dblCount & 1) == 0
+      if (!isEven) {
+        throw new Exception("Triangle count resulted in an invalid number of triangles.")
```
Comment (srowen) [inline graphx/src/main/scala/org/apache/spark/graphx/lib/TriangleCount.scala]: I would just `require(...)` this condition. `(dblCount % 2) == 0` would be a little more canonical

File: graphx/src/main/scala/org/apache/spark/graphx/lib/TriangleCount.scala [hunk 2]
```diff
@@ -27,25 +28,47 @@ import org.apache.spark.graphx._
...
 
   def run[VD: ClassTag, ED: ClassTag](graph: Graph[VD, ED]): Graph[Int, ED] = {
-    // Remove redundant edges
-    val g = graph.groupEdges((a, b) => a).cache()
+    // Transform the edge data something cheap to shuffle and then canonicalize
+    val canonicalGraph = graph.mapEdges(e => true).removeSelfEdges().convertToCanonicalEdges()
+    // Get the triangle counts
+    val counters = runPreCanonicalized(canonicalGraph).vertices
+    // Join them bath with the original graph
+    graph.outerJoinVertices(counters) { (vid, _, optCounter: Option[Int]) =>
```
Comment (srowen) [inline graphx/src/main/scala/org/apache/spark/graphx/lib/TriangleCount.scala]: Nit: just `{ (_, _, optCounter) => ` right?

File: graphx/src/main/scala/org/apache/spark/graphx/lib/TriangleCount.scala [hunk 3]
```diff
@@ -27,25 +28,47 @@ import org.apache.spark.graphx._
  * The algorithm is relatively straightforward and can be computed in three steps:
  *
  * <ul>
- * <li>Compute the set of neighbors for each vertex
- * <li>For each edge compute the intersection of the sets and send the count to both vertices.
+ * <li> Compute the set of neighbors for each vertex
```
Comment (srowen) [inline graphx/src/main/scala/org/apache/spark/graphx/lib/TriangleCount.scala]: Nit: the `<li>` should be closed if you're writing HTML tags

**Comment by insidedctm** [comment]
@srowen good points, I've updated and pushed changes in line with your comments

**Comment by srowen** [comment]
Jenkins, test this please

**Comment by jegonzal** [comment]
This looks good to me. @insidedctm thanks for reviving the PR and @srowen thanks for taking a look at this!  My only minor concern is that it will change the results for people that are using triangle count so we should note this in the change log on the next release.

**Comment by rxin** [comment]
This is great timing given the next version is 2.0.

**Comment by rxin** [comment]
I'm going to merge this in master. Thanks.

**Pull request closed without merging**